specs/dbnode/flush/FlushVersion.tla

```diff
@@ -0,0 +1,253 @@
+---------------------------- MODULE FlushVersion ----------------------------
+EXTENDS Integers, TLC
+
+\* Constants used to limit the potential search space when
+\* running the TLC model checker.
+CONSTANT maxNumWrites
+CONSTANT maxNumFlushes
+CONSTANT maxNumTicks
+
+(***************************************************************************
+\* This specification models the version-based flushing/ticking of series
+\* buffer buckets that is required for the warm/cold write feature. The
+\* specification models the system by running three different processes:
+\*
+\*    1. One process responsible for beginning and ending the flush process.
+\*       This process will begin a flush by looping through all the BucketsInMemory
+\*       and setting their LastSuccessfulFlushVersion to whatever the
+\*       CurrentLastSuccessfulFlushVersion is. It also models both flush
+\*       successes and flush failures.
+\*          Flush Success: Copy all the BucketsInMemory whose
+|*                         LastSuccessfulFlushVersion is smaller than
+\*                         or equal to the CurrentLastSuccessfulFlushVersion (I.E those
+\*                         we could have conceivably known about at flush start time) into
+\*                         the set of PersistedBuckets, and then increment the
+\*                         CurrentLastSuccessfulFlushVersion.
+\*
+\*          Flush Failure: Nothing happens and the CurrentLastSuccessfulFlushVersion stays the same.
+\*
+\*
+\*    2. One process responsible for creating new buckets by issuing "writes". It does
+\*       this by storing them in the BucketsInMemory set, as well as in the
+\*       WrittenBuckets set (so we can keep track of them independentLY of the eviction
+\*       from memory process).
+\*
+\*
+\*    3. One process responsible for the "Ticking" process that will periodically evict
+\*       every bucket from the BucketsInMemory set whose LastSuccessfulFlushVersion is >=
+\*       CurrentLastSuccessfulFlushVersion and also > 0.
+\*
+\*
+\* The specification verifies the system by allowing all of these processes to run
+\* concurrently (as they do in M3DB) and asserting that the invariant that we can
```

Comment:
How does this get run concurrently? I just wonder if this will always run the three processes in the same order or pattern such that certain cases won't be caught.

Reply:
It runs all processes and generates all possible interleaving of actions :D